TWTValidation/JSON Validator/Schema Validators/TWTJSONSchemaObjectValidator.m

```diff
@@ -0,0 +1,184 @@
+    NSMutableArray *dependenciesErrors = outError ? [[NSMutableArray alloc] init] : nil;
+
+    NSSet *keysSet = [NSSet setWithArray:[value allKeys]];
+
+    if (self.maximumPropertyCount || self.minimumPropertyCount) {
+        TWTValidator *countValidator = [[TWTNumberValidator alloc] initWithMinimum:self.minimumPropertyCount maximum:self.maximumPropertyCount];
+        countValidated = [countValidator validateValue:@([value count]) error:&countError];
+    }
+
+    if (self.requiredPropertyKeys && ![self.requiredPropertyKeys isSubsetOfSet:keysSet]) {
+        requiredPropertiesValidated = NO;
+        //        requiredPropertiesError = [NSError twt_validationErrorWithCode:TWTValidationErrorCodeValueNotInSet failingValidator:self value:value localizedDescription:TWTLocalizedString(@"TWTJSONObjectValidator.requiredPropertyNotPresentError")];
+    }
+
+    NSError *error = nil;
+    for (NSString *key in value) {
+        error = nil;
+        BOOL propertyIsDefined = NO;
+        id objectForKey = [value objectForKey:key];
+
+        TWTKeyValuePairValidator *propertyValidator = [self.propertyAndValidators objectForKey:key];
+        if (propertyValidator) {
+            propertyIsDefined = YES;
+            if (![propertyValidator validateValue:objectForKey error:outError ? &error : NULL]) {
+                propertiesValidated = NO;
+                [propertiesErrors addObjectsFromArray:error.twt_underlyingErrors];
+            }
+        }
+
+        for (TWTKeyValuePairValidator *patternValidator in self.patternPropertyValidators) {
+            NSError *regularExpressionError = nil;
+            error = nil;
+            NSRegularExpression *regularExpression = [NSRegularExpression regularExpressionWithPattern:patternValidator.key options:0 error:&regularExpressionError];
+
+            if (!regularExpressionError) {
+                if ([regularExpression numberOfMatchesInString:key options:0 range:NSMakeRange(0, [key length])] > 0) {
+                    propertyIsDefined = YES;
+
+                    if (![patternValidator validateValue:objectForKey error:outError ? &error : NULL]) {
+                        patternPropertiesValidated = NO;
+                        [patternPropertiesErrors addObjectsFromArray:error.twt_underlyingErrors];
+                    };
+                }
+            } // If regular expression is invalid, no properties will match and thus all instances are valid
+        }
+
+        error = nil;
+        if (!propertyIsDefined) {
+            if (![self.additionalPropertiesValidator validateValue:objectForKey error:outError ? &error : NULL]) {
+                additionalPropertiesValidated = NO;
+                [additionalPropertiesErrors addObjectsFromArray:error.twt_underlyingErrors];
+            }
+        }
+
+        // better to check if self.propertyDependencies is nil first?
+        id dependency = self.propertyDependencies[key];
+        if (dependency) {
```

Comment:
Is it better to check if self.propertyDependencies is nil first? I'd still have to check afterward whether one exists for the given key. (Does it make a difference?)